src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -2347,16 +2363,27 @@ fn specialize_one_pattern<'p, 'tcx>(
             // just integers. The only time they should be pointing to memory
             // is when they are subslices of nonzero slices.
             let (alloc, offset, n, ty) = match value.ty.kind {
-                ty::Array(t, n) => match value.val {
-                    ty::ConstKind::Value(ConstValue::ByRef { offset, alloc, .. }) => {
-                        (alloc, offset, n.eval_usize(cx.tcx, cx.param_env), t)
+                ty::Array(t, n) => {
+                    let n = n.eval_usize(cx.tcx, cx.param_env);
+                    match value.val {
+                        ty::ConstKind::Value(ConstValue::ByRef { offset, alloc, .. }) => {
+                            (Cow::Borrowed(alloc), offset, n, t)
+                        }
+                        ty::ConstKind::Value(ConstValue::Scalar(Scalar::Raw { data, .. }))
+                        if n == 0 => {
+                            let align = Align::from_bytes(data as u64).unwrap();
+                            // empty array
+                            (Cow::Owned(Allocation::zst(align)), Size::ZERO, 0, t)
```

Comment:
```suggestion
                            if ctor_wild_subpatterns.len() as u64 == 0 {
                                return Some(PatStack::empty())
                            } else {
                                return None
                            }
```
When `n == 0`, the whole branch will return `(0..0).collect()`, i.e. `Some(PatStack::empty())`. Should we do that here directly instead of crafting an allocation that won't be used ?

Reply:
That's smart. Let's do that

Reply:
github doesn't allow applying suggestions on multiple lines. I added your suggestion manually